Unrelated string changes in objects garble string changes through `changeAt()`

Hi!

I'm seeing unexpected string merging results when I use `changeAt()` to updated a string property in a nested object. More specifically, when I have 2 objects, certain size of regular updates (through `change()`) to the string property on one object affect the `changeAt()` results on the other object. For context, I'm trying to use `changeAt` to achieve "undo" on some previous changes while preserving the changes afterwards. 

Here is the code snippet to reproduce:

```typescript

describe('Automerge undo', () => {
  it('can undo string property in nested objects with changeAt', () => {
    let doc = Automerge.init('00')

    doc = Automerge.change(doc, (d: any) => {
      d.object1 = {color: 'unset'}
      d.object2 = {name: ''}
    })

    doc = Automerge.change(doc, (d: any) => {
      d.object1.color = 'red'
    })
    const afterRed = Automerge.getHeads(doc)

    doc = Automerge.change(doc, (d: any) => {
      d.object1.color = 'white'
    })
    const afterWhite = Automerge.getHeads(doc)

    // change object1 color again
    doc = Automerge.change(doc, (d: any) => {
      d.object1.color = 'blue'
    })
    const afterBlue = Automerge.getHeads(doc)
    assert.deepEqual(doc.object1.color, 'blue')

    // update string property on unrelated object
    doc = Automerge.change(doc, (d: any) => {
      d.object2.name = 'a'.repeat(31)  // 31 is the breaking point
    })

    // revert blue to white
    doc = Automerge.changeAt(doc, afterBlue, (d: any) => {
      d.object1.color = 'white'
    }).newDoc
    assert.deepEqual(doc.object1.color, 'white')

    // revert white to red
    doc = Automerge.changeAt(doc, afterWhite, (d: any) => {
      d.object1.color = 'red'
    }).newDoc
    assert.deepEqual(doc.object1.color, 'red')

    // revert red to unset
    doc = Automerge.changeAt(doc, afterRed, (d: any) => {
      d.object1.color = 'unset'
    }).newDoc
    assert.deepEqual(doc.object1.color, 'unset')
  })
})

```

Test case failed with following error:
```
  Automerge
    1) can undo string property in nested objects with changeAt


  0 passing (9ms)
  1 failing

  1) Automerge
       can undo string property in nested objects with changeAt:

      AssertionError: expected 'unsetredte' to deeply equal 'unset'
      + expected - actual

      -unsetredte
      +unset

```

If the `object2.name` is updated to a string less than 31 characters, the test case will pass. And if the string is longer (like 40 characters), the test case will fail sooner at `assert.deepEqual(doc.object1.color, 'red')` with `AssertionError: expected 'redwhite' to deeply equal 'red'`. Reordering/frequency of the `change()` operations before `changeAt()` can affect the breaking string size, but general results are consistent.

I also tried to reproduce the case with `clone()` and `merge()` but wasn't able to. My guess was when I called `clone()` with different actorID, updated to a different color, and then called `merge()`, the color changes are regarded as conflict and thus keeping one value instead of merging multiple.

Can you help me understand why change size/frequency in one nested string can affect changes in another nested string? Any pointers to how to better decode changes and debug string merging would also be very helpful!

Thanks in advance!

--- a/javascript/test/patches.ts
+++ b/javascript/test/patches.ts
@@ -137,4 +137,31 @@ describe("patches", () => {
       },
     )
   })
+
+  it("should produce correct patches during changeAt", () => {
+    // This test exercises the bug reported in https://github.com/automerge/automerge/issues/951
+    //
+    // The problem was that the patches emitted by changeAt erroenously included
+    // patches for objects that are not visible in the final state of the document
+    // which cause garbled content. This was especially difficult to track down
+    // because it was only triggered by larger patches
+
+    let doc = Automerge.init<{ name?: string; color?: string }>()
+
+    let beginning = Automerge.getHeads(doc)
+
+    doc = Automerge.change(doc, (d: any) => {
+      d.name = "a".repeat(100) // Bug is triggered by more than 100 patches
+    })
+
+    doc = Automerge.changeAt(doc, beginning, (d: any) => {
+      d.color = "red"
+    }).newDoc
+
+    doc = Automerge.changeAt(doc, beginning, (d: any) => {
+      d.color = "unset"
+    }).newDoc
+    // The bug manifested as `doc.color` being "usetred" rather than the expected "unset"
+    assert.deepStrictEqual(doc.color, "unset")
+  })
 })

--- a/rust/automerge/src/automerge.rs
+++ b/rust/automerge/src/automerge.rs
@@ -1641,7 +1641,7 @@ impl Automerge {
             if !visible_objs.contains(&obj.id) {
                 continue;
             }
-            for op in ops.visible(at.clone()) {
+            for op in ops.visible(at.clone()).top_ops() {
                 //if op.visible_at(at.as_ref()) {
                 if let OpType::Make(_) = op.op_type() {
                     visible_objs.insert(op.id.into());

--- a/rust/automerge/tests/text.rs
+++ b/rust/automerge/tests/text.rs
@@ -6,8 +6,8 @@ use automerge::{
     marks::{ExpandMark, Mark},
     patches::TextRepresentation,
     transaction::Transactable,
-    ActorId, AutoCommit, ConcreteTextValue, ObjType, Patch, PatchAction, ReadDoc, ScalarValue,
-    TextEncoding, ROOT,
+    ActorId, AutoCommit, ConcreteTextValue, ObjType, Patch, PatchAction, Prop, ReadDoc,
+    ScalarValue, TextEncoding, Value, ROOT,
 };
 const B: usize = 16;
 
@@ -858,4 +858,141 @@ fn removed_marks_should_not_appear_in_get_marks() {
     assert_eq!(marks.iter().collect::<Vec<_>>(), vec![]);
     assert_eq!(marks.len(), 0);
     assert_eq!(marks.num_marks(), 0);
+}
+
+#[test]
+fn incorrect_patches_produced_when_isolating_and_integrating() {
+    // This test exercises the bug reported in https://github.com/automerge/automerge/issues/951
+    //
+    // The issue was reported as to do with `changeAt`, hence the use of
+    // `isolate` and `integrate` here. But in fact it occurs whenever there are
+    // conflicting changes and more than 100 patches
+
+    // Hard code actor ID to avoid flakes in patch ordering
+    let actor = ActorId::from_str("aaaaaa").unwrap();
+    let mut doc = AutoCommit::new().with_actor(actor);
+
+    let beginning = doc.get_heads();
+
+    // This issue was observed when the number of patches was larger than
+    // 100. This is because if there are more than 100 patches then the
+    // PatchBuilder::new function precomputes a visible path cache
+    const NUM_CHARS_IN_NEW_NAME: usize = 100;
+    let name = doc.put_object(&ROOT, "name", ObjType::Text).unwrap();
+    let new_name = "a".repeat(NUM_CHARS_IN_NEW_NAME);
+    doc.splice_text(&name, 0, 0, &new_name).unwrap();
+
+    // Create a concurrent change at the root
+    doc.isolate(&beginning);
+    let color = doc.put_object(&ROOT, "color", ObjType::Text).unwrap();
+    doc.splice_text(&color, 0, 0, "red").unwrap();
+    doc.integrate();
+    assert_eq!(doc.text(&color).unwrap(), "red");
+
+    // Pop the patches - this is necessary to make the test work
+    let _ignored_patches = doc.diff_incremental();
+
+    // Make sure the document has a clean patch log
+    doc.commit();
+
+    // Create another concurrent change
+    doc.isolate(&beginning);
+    let color = doc.put_object(&ROOT, "color", ObjType::Text).unwrap();
+    doc.splice_text(&color, 0, 0, "unset").unwrap();
+    doc.commit();
+    doc.integrate();
+
+    let patches = doc.diff_incremental();
+
+    // Here we would expect the patches to be something like:
+    //
+    // first patch back to the beginning of doc
+    // * DeleteMap ROOT "color"
+    // * DeleteMap ROOT "name"
+    //
+    // Now insert new items
+    // * PutMap ROOT "color" Text
+    // * PutMap ROOT "name" Text
+    // * SpliceText "name" "aaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    // * SpliceText "color" "unset"
+    //
+    // But what we actually get is
+    //
+    // * DeleteMap ROOT "color"
+    // * DeleteMap ROOT "name"
+    //
+    // Now insert new items
+    // * PutMap ROOT "color" Text
+    // * PutMap ROOT "name" Text
+    // * DeleteSeq "color" 3
+    // * SpliceText "color" "red"
+    // * SpliceText "name" "aaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    // * SpliceText "unset"
+    //
+    // Where it appears that the 'red' property has been inserted into the patch
+    // log in the wrong order (we would at least expect the deletion to come
+    // after the insertion)
+
+    // This assertion works until NUM_CHARS_IN_NEW_NAME is greater than 44
+    assert_eq!(
+        patches,
+        vec![
+            Patch {
+                obj: ROOT,
+                path: vec![],
+                action: PatchAction::DeleteMap {
+                    key: "color".to_string()
+                },
+            },
+            Patch {
+                obj: ROOT,
+                path: vec![],
+                action: PatchAction::DeleteMap {
+                    key: "name".to_string()
+                },
+            },
+            Patch {
+                obj: ROOT,
+                path: vec![],
+                action: PatchAction::PutMap {
+                    key: "color".to_string(),
+                    value: (Value::Object(ObjType::Text), color.clone()),
+                    conflict: true
+                },
+            },
+            Patch {
+                obj: ROOT,
+                path: vec![],
+                action: PatchAction::PutMap {
+                    key: "name".to_string(),
+                    value: (Value::Object(ObjType::Text), name.clone()),
+                    conflict: false
+                },
+            },
+            Patch {
+                obj: name,
+                path: vec![(ROOT, Prop::Map("name".to_string()))],
+                action: PatchAction::SpliceText {
+                    index: 0,
+                    value: ConcreteTextValue::new(
+                        &new_name,
+                        TextRepresentation::String(TextEncoding::UnicodeCodePoint)
+                    ),
+                    marks: None
+                }
+            },
+            Patch {
+                obj: color,
+                path: vec![(ROOT, Prop::Map("color".to_string()))],
+                action: PatchAction::SpliceText {
+                    index: 0,
+                    value: ConcreteTextValue::new(
+                        "unset",
+                        TextRepresentation::String(TextEncoding::UnicodeCodePoint)
+                    ),
+                    marks: None
+                }
+            }
+        ]
+    );
 }

